sdk/python/lib/pulumi/provider/experimental/analyzer.py

```diff
@@ -85,19 +92,174 @@ def analyze_component(
         args = component.__init__.__annotations__.get("args", None)
         if not args:
             raise Exception(
-                f"Could not find `args` keyword argument in {component}'s __init__ method"
+                f"ComponentResource '{component.__name__}' requires an argument named 'args' with a type annotation in its __init__ method"
             )
         return ComponentDefinition(
             description=component.__doc__.strip() if component.__doc__ else None,
-            inputs=self.analyze_types(args),
-            outputs=self.analyze_types(component),
+            inputs=self.analyze_type(args),
+            outputs=self.analyze_type(component),
         )
 
-    def analyze_types(self, typ: type) -> dict[str, PropertyDefinition]:
-        return {}
+    def analyze_type(self, typ: type) -> dict[str, PropertyDefinition]:
+        """
+        analyze_types returns a dictionary of the properties of a type based on
+        its annotations.
+
+        For example for the class
+
+            class SelfSignedCertificateArgs:
+                algorithm: pulumi.Output[str]
+                bits: Optional[pulumi.Output[int]]
+
+        we get the following properties:
+
+            {
+                "algorithm": SchemaProperty(type=PropertyType.STRING, optional=False),
+                "bits": SchemaProperty(type=PropertyType.INTEGER, optional=True)
+            }
+        """
+        if not hasattr(typ, "__annotations__"):
+            return {}
+        return {k: self.analyze_property(v) for k, v in typ.__annotations__.items()}
+
+    def analyze_property(
+        self, arg: type, optional: Optional[bool] = None
+    ) -> PropertyDefinition:
+        """
+        analyze_arg analyzes a single annotation and turns it into a SchemaProperty.
+        """
+        optional = optional if optional is not None else is_optional(arg)
+        unwrapped = None
+        ref = None
+        if is_plain(arg):
+            unwrapped = arg
```

Comment:
In a later PR we'll have to make a distinction between plain and inputty/outputty properties and reflect that in the resulting schema. Worth adding a TODO?

Reply:
Yes, done